snuba/datasets/entities/factory.py

```diff
@@ -0,0 +1,57 @@
+from typing import Callable, MutableMapping, Set, Sequence
+
+from snuba import settings
+from snuba.datasets.entity import Entity
+from snuba.util import with_span
+
+ENTITY_IMPL: MutableMapping[str, Entity] = {}
+ENTITY_NAME_LOOKUP: MutableMapping[Entity, str] = {}
+
+ENTITY_NAMES: Set[str] = {
+    "outcomes",
+    "outcomes_raw",
+}
+
+ACTIVE_ENTITY_NAMES: Set[str] = ENTITY_NAMES - settings.DISABLED_ENTITIES
+
+
+class InvalidEntityError(Exception):
+    """Exception raised on invalid entity access."""
+
+
+@with_span()
+def get_entity(name: str) -> Entity:
+    if name in ENTITY_IMPL:
+        return ENTITY_IMPL[name]
+
+    if name in settings.DISABLED_ENTITIES:
+        raise InvalidEntityError(
+            f"entity {name!r} is not available in this environment"
+        )
+
+    from snuba.datasets.entities.outcomes import OutcomesEntity
+    from snuba.datasets.entities.outcomes_raw import OutcomesRawEntity
+
+    entity_factories: MutableMapping[str, Callable[[], Entity]] = {
+        "outcomes": OutcomesEntity,
+        "outcomes_raw": OutcomesRawEntity,
+    }
+
+    try:
+        entity = ENTITY_IMPL[name] = entity_factories[name]()
+        ENTITY_NAME_LOOKUP[entity] = name
+    except KeyError as error:
+        raise InvalidEntityError(f"entity {name!r} does not exist") from error
+
+    return entity
+
+
+def get_entity_name(entity: Entity) -> str:
```

Comment:
Wouldn't it be easier for an entity to just know its own name via `entity_factories[name](name)` instead of a top-level reverse dict?

Reply:
I'm going to revisit this, I copied this verbatim from the dataset factory.

Reply:
So I can go either way here. I can add the name to every class, or leave that logic here. I don't think it matters either way.